Add the Context notion. (#424)
open-telemetry/opentelemetry-specification: specification/context.md

# Context

<details>
<summary>
Table of Contents
</summary>

- [Overview](#overview)
- [Create a key](#create-a-key)
- [Get value](#get-value)
- [Set value](#set-value)
- [Optional operations](#optional-operations)
    - [Get current Context](#get-current-context)
    - [Attach Context](#attach-context)
    - [Detach Context](#detach-context)

</details>

## Overview

> `Context` is a propagation mechanism which carries execution-scoped values

Reviewer: ```suggestion
A `Context` is a propagation mechanism which carries execution-scoped values
```

across API boundaries and between logically associated execution units.
Cross-cutting concerns access their data in-process using the same shared
`Context` object.

> `Context` MUST be immutable, and its write operations MUST

Reviewer: ```suggestion
A `Context` MUST be immutable, and its write operations MUST
```

> result in the creation of a new `Context` cointaining the original

Reviewer: ```suggestion
result in the creation of a new `Context` containing the original
```

values and the specified values updated.

Languages are expected to use the single, widely used `Context` implementation
if one exists for them. In the cases where an extremely clear, pre-existing
option is not available, OpenTelemetry MUST provide its own `Context`
implementation. Depending on the language, its usage may be either explicit
or implicit.

Users writing instrumentation in languages that use `Context` implicitly are
discouraged from using the `Context` API directly. In those cases, users will
manipulate `Context` through cross-cutting concerns APIs instead, in order to
perform operations such as setting trace or correlation context values for
a specified `Context`.

> `Context` is expected to have the following operations, with their

Reviewer: ```suggestion
A `Context` is expected to have the following operations, with their
```

The API MUST return an opaque object representing the newly created key.

## Get value

Concerns can access their local state in the current execution state
represented by a `Context`.

The API MUST accept the following parameters:

- The `Context`.
- The key.

The API MUST return the value in the `Context` for the specified key.

## Set value

Concerns can record their local state in the current execution state
represented by a `Context`.

The API MUST accept the following parameters:

- The `Context`.
- The key.
- The value to be set.

The API MUST return a new `Context` containing the new value.

## Optional Global operations

These operations are expected to only be implemented by languages
using `Context` implicitly, and thus are optional. These operations
SHOULD only be used to implement automatic scope switching and define
higher level APIs by SDK components and OpenTelemetry instrumentation libraries.

### Get current Context

The API MUST return the `Context` associated with the caller's current execution unit.

### Attach Context

Associates a `Context` with the caller's current execution unit.

The API MUST accept the following parameters:

- The `Context`.

> The API MUST return an object that can be used as `Token` to restore the previous

Reviewer: ```suggestion
The API MUST return an object that can be used as a `Token` to restore the previous
```

# Context

<details>
<summary>
Table of Contents
</summary>

- [Get value](#get-value)
- [Set value](#set-value)
- [Remove value](#remove-value)
- [Optional operations](#optional-operations)
    - [Get current Context](#get-current-context)
    - [Set current Context](#set-current-context)

</details>

> `Context` is a propagation mechanism which carries execution-scoped-values

Reviewer 2: ```suggestion
`Context` is a propagation mechanism which carries execution-scoped values
```

> across API boundaries and between execution units.

Reviewer 2: If it is execution-scoped (what does that even mean? -- I think I know but I think we need a more clear definition, probably with examples), then how can it carry between execution units? Should it not rather have separate values between them? Maybe:

```suggestion
across API boundaries and between logically associated execution units.
```

But that adds even more vagueness.

Author: > how can it carry between execution units?

The idea is that you can pass a `Context` between execution units (threads or greenlets, etc). I will add a little bit of clarification there.

Languages are expected to use the single, widely used `Context` implementation
if one exists for them. In the cases where an extremely clear, pre-existing
option is not available, OpenTelemetry should provide its own `Context`
implementation. Depending on the language, its usage may be either explicit
or implicit.

> `Context` is meant to primarily be a SDK API, and thus be used by SDK and

Reviewer 3: This is clearly not the case in Go.

Author: Right, good call. I need to add a note here on the fact this is depends on the language.

cross-cutting concerns authors, rather than users writing instrumentation.

`Context` is expected to have the following operations, with their
respective language differences:

#### Get value

Concerns can access their local state out of a specified `Context`.

The API SHOULD accept the following parameters:

- (Required) The `Context`.
- (Required) The concern identifier.

The API SHOULD return the value in the `Context` for the specified concern
identifier.

#### Set value

> Concerns can record their local state in a specified `Context`.

Reviewer 3: Context is immutable, let's find more accurate wording. Eg "current execution state represented by a Context".

Reviewer 2: Should this be up to the language?

Reviewer 2: But if we say immutable, then such a function would typically be called `WithValue`.

The API SHOULD accept the following parameters:

- (Required) The `Context`.
- (Required) The concern identifier.
- (Required) The value to be set.

The API SHOULD return a new `Context` containing the new value.
The new value will not be present in the old `Context`.

#### Remove value

Concerns can clear their local state in a specified `Context`.

The API SHOULD accept the following parameters:

- (Required) The `Context`.
- (Required) The concern identifier.

The API SHOULD return a new `Context` with the value cleared.
The removed value still remains present in the old `Context`.

### Optional Global operations

> These operations are optional, and are protected to be specifically

Reviewer 3: What does "protected" mean?

Author: Oh, this means that end users shouldn't be touching (abusing) these calls ;) Will add a note on this, and we can discuss it from there.

> These operations are optional, and are protected to be specifically

Reviewer 2: ```suggestion
These operations are optional, and are protected to be specifically used to
```

or

```suggestion
These operations are optional, and are protected to specifically
```

implement automatic scope switching and define higher level APIs.

#### Get current Context

> The API SHOULD return the `Context` associated with program execution.

Reviewer 2: ```suggestion
The API SHOULD return the `Context` associated with the caller's current execution unit.
```

Use consistent terminology.

#### Set current Context

> Associates a `Context` with program execution.

Reviewer 2: ```suggestion
Associates a `Context` with the caller's current execution unit.
```

# Context

<details>
<summary>
Table of Contents
</summary>

- [Overview](#overview)
- [Get value](#get-value)
- [Set value](#set-value)
- [Remove value](#remove-value)
- [Optional operations](#optional-operations)
    - [Get current Context](#get-current-context)
    - [Set current Context](#set-current-context)

</details>

## Overview

`Context` is a propagation mechanism which carries execution-scoped values
across API boundaries and between logically associated execution units.
Cross-cutting concerns access their data in-process using the same shared
`Context` object.

Languages are expected to use the single, widely used `Context` implementation
if one exists for them. In the cases where an extremely clear, pre-existing
option is not available, OpenTelemetry MUST provide its own `Context`
implementation. Depending on the language, its usage may be either explicit
or implicit.

Users writing instrumentation in languages that implicitly use `Context`
are discouraged to use the `Context` API directly.

`Context` is expected to have the following operations, with their
respective language differences:

> ## Get value

Reviewer 4: Are these methods API methods or SDK methods?

Author: None, if you mean whether they are OTel API vs SDK.

About `Context` itself, they are expected to be fully implemented, and they are expected to be mostly used by the OTel SDK (as mentioned in the overview of this document). Was this what wanted to know?

Author: Hey @dyladan I appended this small comment at the Overview:

> `Context` can be considered as an implementation detail of the SDK.
Users will manipulate `Context` through the cross-cutting concerns APIs
rather than directly.

Resolving this comment. Feel free to provide on the up-to-date doc in case you think we need further clarification.

Concerns can access their local state in the current execution state
represented by a `Context`.

The API SHOULD accept the following parameters:

- (Required) The `Context`.
- (Required) The concern identifier.

The API MUST return the value in the `Context` for the specified concern
identifier.

## Set value

Concerns can record their local state in the current execution state
represented by a `Context`.

The API SHOULD accept the following parameters:

- (Required) The `Context`.
- (Required) The concern identifier.
- (Required) The value to be set.

The API SHOULD return a new `Context` containing the new value.
The `Context` passed as parameter SHOULD NOT be modified.

## Remove value

Concerns can clear their local state in a specified `Context`.

The API SHOULD accept the following parameters:

- (Required) The `Context`.
- (Required) The concern identifier.

The API MUST return a new `Context` with the value cleared.
The `Context` passed as parameter MUST NOT be modified.

## Optional Global operations

> These operations are optional, and SHOULD ONLY be used to

Reviewer 4: What is meant by optional? Does the API decide to have them or not, or does the API need to support SDK implementations which may or may not have this?

Author: This is mostly tied to the language - go will not have them, as its `Context` handling is explicit - but Python or Java do expose these operations, as they have implicit handling ;)

# Context

<details>
<summary>
Table of Contents
</summary>

- [Overview](#overview)
- [Get value](#get-value)
- [Set value](#set-value)
- [Remove value](#remove-value)
- [Optional operations](#optional-operations)
    - [Get current Context](#get-current-context)
    - [Set current Context](#set-current-context)

</details>

> ## Overview

Reviewer 2: What seems to be missing here, is an API to create a new context.

Author: This was left out in purpose, to simplify things. If you really feel we should include it, I'd suggest working on this after this PR has been merged (I know a few people really dislike the idea of having it here, but we could discuss it nevertheless).

Reviewer 5: If the API for creating contexts needs to be mentioned, I think it should go along with tools for explicit context management: create an empty context, switch from one context to another, and get the current context for later use.

Reviewer 2: Without an API to create a context, its impossible to implement a propagator.
And given that 3rd party SDKs should also use the same Context, they need this to even create a Span.

Reviewer 4: It seems to me that there is (or should be) an implicit 'root' context which is empty if no context has been set.

Author: In my opinion both the root instance and a `Context` creation step are implementation details, so to speak - i.e. too low level.

I'm afraid that, if we end up adding items like this, this will suddenly explode to have 5-6 more operations, that will be deemed as important. 

That being said, I'd like to hear other thoughts on this.

`Context` is a propagation mechanism which carries execution-scoped values
across API boundaries and between logically associated execution units.
Cross-cutting concerns access their data in-process using the same shared
`Context` object.

Languages are expected to use the single, widely used `Context` implementation
if one exists for them. In the cases where an extremely clear, pre-existing

> option is not available, OpenTelemetry should provide its own `Context`

Author: Let's go with `MUST` for now. So far only Ruby had to implement its own, and we will see later how it goes (and how it affects this spec).

implementation. Depending on the language, its usage may be either explicit
or implicit.

> For languages without explicit `Context` usage, this should

Author: I'd prefer to have it as: 

> Users writing instrumentation in languages that implicitly use Context should are discouraged to use the Context API directly.

Would that work for you?

> For languages without explicit `Context` usage, this should

Reviewer 6: I was mostly asking so I could get a better understanding of what is trying to be said here.

Based on the rephrasing(s), I'm not sure the sentence needs to be included in the specification.
It isn't a normative requirement and it really doesn't seem to add actionable value to the reader.

I would recommend just removing it.

> For languages without explicit `Context` usage, this should

Author: We really want to prevent users from abusing this API. This is supposed to be a rather advanced feature, and I'd rather keep it this way. Better be safe than sorry ;) will update it with the latest rephrasing :)

be considered to primarily be a SDK API, and thus be used by SDK and
cross-cutting concerns authors, rather than users writing instrumentation.

`Context` is expected to have the following operations, with their
respective language differences:

## Get value

Concerns can access their local state in the current execution state
represented by a `Context`.

> The API SHOULD accept the following parameters:

Author: This is a rather abstract part and I'd prefer to use SHOULD for flexibility purposes. Remember: we don't control all the implementations, and this is a more of a overall-expected API.

> The API SHOULD accept the following parameters:

Reviewer 6: Gotcha. If that is the desire than the parameters should not be prefixed with `(Required)`.

- (Required) The `Context`.
- (Required) The concern identifier.

> The API SHOULD return the value in the `Context` for the specified concern

Author: This one makes sense to use `MUST`, yes.

identifier.

## Set value

Concerns can record their local state in the current execution state
represented by a `Context`.

The API SHOULD accept the following parameters:

- (Required) The `Context`.
- (Required) The concern identifier.
- (Required) The value to be set.

The API SHOULD return a new `Context` containing the new value.
The `Context` passed as parameter SHOULD NOT be modified.

## Remove value

Concerns can clear their local state in a specified `Context`.

The API SHOULD accept the following parameters:

- (Required) The `Context`.
- (Required) The concern identifier.

The API SHOULD return a new `Context` with the value cleared.

> The `Context` passed as parameter SHOULD NOT be modified.

Author: Again, not something we can fully control.

> The `Context` passed as parameter SHOULD NOT be modified.

Reviewer 6: Oh! Are we saying that the ***user who passed the argument*** should not modify the value, not that the SDK/API should not modify the value?

> The `Context` passed as parameter SHOULD NOT be modified.

Author: So lets go with `MUST` for now, while keeping SHOULD for flexibility purposes in the arguments. If/when an implementation updating `Context` in place appears, we can revisit this one.

# Context

<details>
<summary>
Table of Contents
</summary>

- [Overview](#overview)
- [Get value](#get-value)
- [Set value](#set-value)
- [Remove value](#remove-value)
- [Optional operations](#optional-operations)
    - [Get current Context](#get-current-context)
    - [Set current Context](#set-current-context)

</details>

## Overview

`Context` is a propagation mechanism which carries execution-scoped values
across API boundaries and between logically associated execution units.
Cross-cutting concerns access their data in-process using the same shared
`Context` object.

> `Context` MUST be immutable, and all its operations MUST

Reviewer 7: Naive question: what are the advantages of having an immutable context?

Reviewer 4: It is important in multithreaded scenarios. If you have an incoming request (single context) which then spawns two threads of execution, if the context is mutable then a single thread of execution can't be sure that another thread is not modifying its context. If changing context requires you to create a new context, then by definition any thread of execution outside of yours can only create new contexts which you don't worry about and cannot modify yours.

1. Thread A accepts an incoming request, and creates a context c1.
2. Thread A begins processing the request with two threads a1 and a2. Because context has not been modified, they both have context c1.
3. Thread a2 updates context by creating a new context c2. Thread a1 still has context c1 which remains unchanged and everyone is happy.

Author: Also, most of the `Context` implementations (all?) have this invariant, for the reasons @dyladan exposed :)

Reviewer 2: Also, "copying" an immutable context can be implemented very efficiently, e.g. see Python's contextvars implementation https://www.python.org/dev/peps/pep-0550/#implementation which is based on a ["hash array map trie"](https://en.wikipedia.org/wiki/Hash_array_mapped_trie)

# Context

<details>
<summary>
Table of Contents
</summary>

- [Overview](#overview)
- [Get value](#get-value)
- [Set value](#set-value)
- [Remove value](#remove-value)
- [Optional operations](#optional-operations)
    - [Get current Context](#get-current-context)
    - [Set current Context](#set-current-context)

</details>

## Overview

`Context` is a propagation mechanism which carries execution-scoped values
across API boundaries and between logically associated execution units.
Cross-cutting concerns access their data in-process using the same shared
`Context` object.

Languages are expected to use the single, widely used `Context` implementation
if one exists for them. In the cases where an extremely clear, pre-existing
option is not available, OpenTelemetry MUST provide its own `Context`
implementation. Depending on the language, its usage may be either explicit
or implicit.

> `Context` can be considered as an implementation detail of the SDK.

Reviewer 3: This phrase is misleading.  In languages with explicit context (like Go), the API is directly exposed to Context, so it's not an implementation detail. The following sentence is good.

Users will manipulate `Context` through the cross-cutting concerns APIs
rather than directly.

Users writing instrumentation in languages that

> use `Context` implicitly are discouraged to use the `Context` API directly.

Reviewer 3: ```suggestion
use `Context` implicitly are discouraged from using the `Context` API directly.
```

In those cases, users will manipulate `Context` through the cross-cutting

> concerns APIs instead.

Reviewer 8: nit: indentation?

> concerns APIs instead.

Reviewer 8: Not sure I understand this sentence, can you clarify a bit?

Author: Added a small note.

`Context` is expected to have the following operations, with their

> respective language differences:

Reviewer 8: We need to have a notion of a `Key` that is retrieved from the `Context` and all operations work against that `Key`. The main reason is that the element in the `Context` should be identified by the `Key` and not directly a `String` to avoid string names collisions. See what grpc.Context does, go Context does the same also python `ContextVar` is the same.

Author: I decided to leave it out for now, in order to simplify this PR (in a _similar_ spirit to leaving "Remove value" out for now).

I personally think this is an improvement, and personally would leave this out for now, but can add it back (as I originally had). If it becomes too polemical for some reason, I will definitely remove it again and work on it for 0.4.0 ;)

Author: Restored. Please review.

Concerns can access their local state in the current execution state
represented by a `Context`.

The API MUST accept the following parameters:

- The `Context`.
- The concern identifier.

The API MUST return the value in the `Context` for the specified concern
identifier.

## Set value

Concerns can record their local state in the current execution state
represented by a `Context`.

The API MUST accept the following parameters:

- The `Context`.
- The concern identifier.
- The value to be set.

> The API MUST return a new `Context` containing the new value.

Reviewer 3: I suggest clarifying what happens to _other_ keys/values, namely that they are still accessible through the returned new instance of the Context. Implementations of the Context may implement it by chaining the contexts, or by cloning the contents.

The `Context` passed as parameter MUST NOT be modified.

## Remove value

Concerns can clear their local state in a specified `Context`.

The API MUST accept the following parameters:

- The `Context`.
- The concern identifier.

The API MUST return a new `Context` with the value cleared.

> The `Context` passed as parameter MUST NOT be modified.

Reviewer 3: This is repeated several times. Why don't we state at the beginning that a Context must be immutable?

> The API MUST return a new `Context` with the value cleared.

Reviewer 8: I checked some implementations go, python, java (grpc) and none of them have the remove, can we not add this right now?

Author: That works, yes. Lets do a follow up on that if/when needed.

Author: Done.

## Optional Global operations

These operations are optional, and SHOULD only be used to
implement automatic scope switching and define higher level APIs

> by SDK components and OpenTelemetry instrumentation libraries.

Reviewer 8: Comment that this are not available for manually propagated Context like Go.

Author: Added a minor note there.

### Get current Context

The API MUST return the `Context` associated with the caller's current execution unit.

### Set current Context

Associates a `Context` with the caller's current execution unit.

The API MUST accept the following parameters:

> - The `Context`.

Reviewer 8: We should encourage to have this return a Token that is used to restore the Context. See Java (grpc) and https://docs.python.org/3/library/contextvars.html#contextvars.ContextVar.set which does the same at a value level because set makes that value current.

Author: Added another minor note as well.

Reviewer 8: We probably need a restore where we use the handle to restore to the previous Context.

PS: attach/detach are probably better names for these 2 methods.

Author: I'd leave Restore out for now as well (for simplicity purposes - lets add it when we add "Remove" from `Context` to have a more complete set of operations).

And you mean:

Set current Context -> Attach Context
Restore Context -> Detach Context

?

To me that sounds great, but _probably_ too Java-ish ;)

Reviewer 8: Restore/Detach is not equivalent to Remove. You need to have code like this:

```java
Token token = Context.attach(newContext);
try {
  // your code;
} finally {
  Context.detach(token);
}
```

So detach is very important compared with Remove, I don't see how a Context would work without detach, because you will set a new Context and that will stay active even after you are out of scope. Another option is to have `Context.attach(newContext)` returning a Scope that when it goes out of scope restores the previous Context.

Reviewer 8: What I am trying to say is that we have two options to achieve this, we either return a `Scope` object that handles the detach or we return a `Token` that can be manually used to call `detach`. The `Scope` API is easier to use and cleaner, but a bit more limited so maybe having both is the right way to do it, you would have primitives as well as a high level easier to use in 99% of the cases.

Reviewer 7: What are the benefits of returning a handle to restore the context?

Author: @bogdandrutu Decided, for simplicity, to go with the `Token` approach - the second one sounds like sugar on top, and while it's nice to have it, we can skip it for now IHMO.

@mauriciovasquezbernal Hiding details overall, and make it simpler to restore things. I know `contextvars` [has them](https://docs.python.org/3/library/contextvars.html#contextvars.ContextVar.reset), for example. But @codeboten also has doubts about this. Would you guys mind chiming it?

Author: By the way, added a small section on the `Detach Context` part. Please review.

Reviewer 9: My question around this change is that since the Context API already provides a mechanism to explicitly set and get a Context, how will using a token to restore a previous Context differ from getting a previous Context then setting it again when wanting to restore?

Reviewer 8: @codeboten this is a very interesting point about why not just simply having one `set` is not enough, because as you said it can be used to restore the previous Context.

There are few reasons:
1) Checking correctness - by having a restore that accepts a Token or just simply the previous Context you can check if the given value is the expected one (in other words you can see the calls to `set` as push to a stack and `restore` as pop where you check that the element you intend to pop is the expected one. This is critical and I've seen cases where auth tokens were leaked because of this issue (I don't think I can share too many things about this issue, but the root cause were people calling set without a paired restore).
2) This approach guarantees (at least tries to enforce because mistakes can still happen) that the Context is not "modified" by calling a function:
```java
void bar() {
  Context current = Context.current();
  foo();
  // This should always be the case, otherwise for examples
  // credentials can be leaked here.
  assert(current == Context.current());
}
``` 
It can still happen if a Set is not followed by a Restore, but by having the Restore users can identify very soon this wrong behavior because the next Restore call will give a signal to the users (log, crash, based on a config).

So this API helps users use the Context API in an expected way.

Reviewer 10: Can this reasoning be added to the spec? It became a topic for discussion without any sort of information in the merged spec itself to help substantiate: https://github.com/open-telemetry/opentelemetry-python/pull/429#discussion_r383989943

Author: Added https://github.com/open-telemetry/opentelemetry-specification/issues/490

> - (Required) The `Context`.

Reviewer 6: Guessing this was just missed, but do we want to remove the `(Required)` prefix to remain consistent with the other parameters sections?

Author: Yeah, totally missed this one.

`Context` object.

> `Context` MUST be immutable, and all its operations MUST

Reviewer 4: This implies that even `get value` would create a new context which I don't believe is the intention.

Author: Oh, good call.

The API MUST return the `Context` associated with the caller's current execution unit.

### Attach Context

Associates a `Context` with the caller's current execution unit.

The API MUST accept the following parameters:

- The `Context`.

> The API MUST return an object that can be used as `Token` to restore the previous

Reviewer 4: I think it would suffice to just say `value` rather than `object` as it may be a string, symbol, or other non-object value

Reviewer 8: It will be good to have in the Token the current and previous Contexts to check that you are restoring the expected Context.
